### PR TITLE
[routing-manager] determine route preference from parent link quality

### PR DESCRIPTION
**Pull request by abtink**
This commit updates how `RoutePublisher` determines the published route preference:
- Medium preference is used on a border router (BR) acting as a router or a BR acting as an end-device (ED) connected to a parent with link quality 1.
- An ED BR with lower link quality publishes its route with low preference.

For ED BRs, the preference is updated if the parent's link quality changes. However, to avoid frequent preference changes due to link variability, the following rules are used:
- If the link quality goes to 2 or 3, the route preference is immediately changed to low.
- On transition to link quality 1, we wait for 5 minutes before changing the preference to medium.
- If the ED BR switches parents, the link quality to the new parent is used to determine preference.

It also adds new APIs and related CLI commands to allow users to get and manually set the published route preference. A new test script `test-021-br-route-prf.py` is added to validate the selection of route preference by BR.

This commit also add a new mechanism to detect when the link quality to the parent changes and signal it using a newly added `Notifier` event.

----

Related to [SPEC-1190](https://threadgroup.atlassian.net/browse/SPEC-1190).

### CODE REVIEW

File: include/openthread/instance.h [hunk 1]
```diff
@@ -171,44 +171,38 @@ uint64_t otInstanceGetUptime(otInstance *aInstance);
...
+#define OT_CHANGED_SUPPORTED_CHANNEL_MASK (1U << 22)       ///< Supported channel mask changed
+#define OT_CHANGED_COMMISSIONER_STATE (1U << 23)           ///< Commissioner state changed
+#define OT_CHANGED_THREAD_NETIF_STATE (1U << 24)           ///< Thread network interface state changed
+#define OT_CHANGED_THREAD_BACKBONE_ROUTER_STATE (1U << 25) ///< Backbone Router state changed
+#define OT_CHANGED_THREAD_BACKBONE_ROUTER_LOCAL (1U << 26) ///< Local Backbone Router configuration changed
+#define OT_CHANGED_JOINER_STATE (1U << 27)                 ///< Joiner state changed
+#define OT_CHANGED_ACTIVE_DATASET (1U << 28)               ///< Active Operational Dataset changed
+#define OT_CHANGED_PENDING_DATASET (1U << 29)              ///< Pending Operational Dataset changed
+#define OT_CHANGED_NAT64_TRANSLATOR_STATE (1U << 30)       ///< The state of NAT64 translator changed
+#define OT_CHANGED_PARENT_LINK_QUALITY (1U << 31)          ///< Parent link quality changed
```
Comment (abtink) [inline include/openthread/instance.h:205]: We change this from `enum` to `#define` so to be able to use the bit `(1U << 31)`. Toolchains do not allow C `enum` constants that won't fit in an `int` type (which can be an `int32_t` on some platforms and therefore won't allow `(1 << 31`) as a valid value)

File: src/core/border_router/routing_manager.cpp [hunk 2]
```diff
@@ -2765,18 +2767,47 @@ void RoutingManager::RoutePublisher::ClearPreference(void)
...
+            UpdatePreference(NetworkData::kRoutePreferenceLow);
+            mTimer.Stop();
+        }
+    }
+
+exit:
+    return;
 }
 
+void RoutingManager::RoutePublisher::HandleTimer(void) { UpdatePreference(NetworkData::kRoutePreferenceMedium); }
```
Comment (EskoDijk) [inline src/core/border_router/routing_manager.cpp:2809]:
```suggestion
void RoutingManager::RoutePublisher::HandleTimer(void) { SetPreferenceBasedOnRole(); }
```

Ideally this would call `SetPreferenceBasedOnRole()` , because there could be code there (not now - but in the future) that determines the preference based on the current role. And current role still needs to be taken into account for the determination after the timer expires.

Comment (abtink) [reply src/core/border_router/routing_manager.cpp]: Changed this in new push.

**Comment by abtink** [comment]
New push contains the (squashed) commit from (now closed) PR:
- https://github.com/openthread/openthread/pull/9094

File: src/cli/cli_br.cpp [hunk 3]
```diff
@@ -44,6 +44,32 @@
 namespace ot {
 namespace Cli {
 
+/**
+ * @cli br init
+ * @code
+ * br init 2 1
+ * Done
+ * @endcode
+ * @cparam br init @ca{if-index} @ca{is-running}
```
Comment (Vyrastas) [inline src/cli/cli_br.cpp:53]: Can we elaborate a little on what the if-index parameter is?  Or maybe spell it out interface-index

For myself, I have no clue where that comes from (how to identify the index).  The API reference doesn't explain it either.

File: include/openthread/border_routing.h [hunk 4]
```diff
@@ -213,6 +213,44 @@ void otBorderRoutingSetRouteInfoOptionPreference(otInstance *aInstance, otRouteP
...
+ * @returns The current published route preference.
+ *
+ */
+otRoutePreference otBorderRoutingGetRoutePreference(otInstance *aInstance);
+
+/**
+ * Explicitly sets the preference of published routes in Network Data.
+ *
+ * After a call to this function, BR will use the given preference. The preference can be cleared by calling
+ * `otBorderRoutingClearRoutePreference`()`.
```
Comment (Vyrastas) [inline include/openthread/border_routing.h:235]: Extra ` between e and (

**Review by Vyrastas — CHANGES_REQUESTED** [review]

File: src/cli/cli_br.cpp (hunk 3 above)
Comment (abtink) [reply src/cli/cli_br.cpp]: Updated it to use `infrastructure-interface-index`.

This is the netif index for the adjacent infrastructure link, which is the infra netif on which we act as a Border Router (BR).

In my view, this API would have been better as a platform API rather than a public OpenThread API. Defining it in `platform/infra_if.h` would have been a better fit, as it already matches the existing `otPlatInfraIfStateChanged()` which is used to notify when the infra netif state changes. However, we generally try to avoid changing public APIs (to avoid compatibility issues), so it is tricky to change this now.

Ideally, an OT user should not need to use the `otBorderRoutingInit()` API. It should be called by the platform layer. I added it to the CLI in this PR only to be able to use it in the test-case script.

File: src/cli/README_BR.md [hunk 5]
```diff
@@ -206,3 +208,37 @@ Clear a previously set preference value for advertising Route Info Options (e.g.
...
+```bash
+> br routeprf low
+Done
+```
+
+### routeprf clear
+
+Usage: `br routeprf clear`
+
+Clear a previously set preference value for publishing routes in Thread Network Data.When cleared BR will automatically determine the route preference based on device's role and link quality to parent (when acting as end-device).
```
[the remark this answers is not in the document]
Comment (abtink) [reply src/cli/README_BR.md]: Updated in new push.

File: include/openthread/border_routing.h (hunk 4 above)
Comment (abtink) [reply include/openthread/border_routing.h]: Thanks. Updated in new push.

File: src/cli/README_BR.md [hunk 6]
```diff
@@ -206,3 +208,37 @@ Clear a previously set preference value for advertising Route Info Options (e.g.
 > br rioprf clear
 Done
 ```
+
+### routeprf
+
+Usage: `br routeprf`
+
+Get the preference used for publishing routes in Thread Network Data.
```
Comment (EskoDijk) [inline src/cli/README_BR.md:216]:
```suggestion
Get the preference used for publishing routes in Thread Network Data. This may be the automatically determined route preference, or an administratively set fixed route preference - if applicable.
```

To clarify what is exactly returned here.

File: src/cli/README_BR.md [hunk 7]
```diff
@@ -206,3 +208,37 @@ Clear a previously set preference value for advertising Route Info Options (e.g.
...
+> br routeprf
+med
+Done
+```
+
+### routeprf \<prf\>
+
+Usage: `br routeprf high|med|low`
+
+Set the preference (which may be 'high', 'med', or 'low') to use publishing routes in Thread Network Data.
```
Comment (EskoDijk) [inline src/cli/README_BR.md:228]:
```suggestion
Set the preference (which may be 'high', 'med', or 'low') to use publishing routes in Thread Network Data. Setting a preference value overrides the automatic route preference determination. It is used only for an explicit administrative configuration of a Border Router.
```

Add some more context and link to spec language "administrative".

Comment (abtink) [reply src/cli/README_BR.md]: Thanks. Updated in new push.

File: src/cli/cli_br.cpp (hunk 3 above)
Comment (Vyrastas) [reply src/cli/cli_br.cpp]: Sounds good, thanks for the explanation.

**Review by Vyrastas — APPROVED** [review]

File: examples/platforms/simulation/infra_if.c [hunk 8]
```diff
@@ -49,13 +49,13 @@ otError otPlatInfraIfSendIcmp6Nd(uint32_t            aInfraIfIndex,
     OT_UNUSED_VARIABLE(aBuffer);
     OT_UNUSED_VARIABLE(aBufferLength);
 
-    return OT_ERROR_FAILED;
+    return OT_ERROR_NONE;
```
Comment (jwhui) [inline examples/platforms/simulation/infra_if.c:52]: Is this change intentional?

Comment (abtink) [reply examples/platforms/simulation/infra_if.c:52]: Yes. So we can emulate a `infra_if` platform (which does nothing) under the simulation-platform. This will allow us to run the new `test-021-br-route-prf.py`.

File: include/openthread/instance.h [hunk 9]
```diff
@@ -53,7 +53,7 @@ extern "C" {
  * @note This number versions both OpenThread platform and user APIs.
  *
  */
-#define OPENTHREAD_API_VERSION (335)
+#define OPENTHREAD_API_VERSION (336)
```
Comment (jwhui) [inline include/openthread/instance.h:56]: Need to rebase and increment :)

**Review by jwhui — APPROVED** [review]
